### PR TITLE
Package for PTL installation

### PR DESCRIPTION
**Pull request by sandisamp**
#### Bug/feature Description
* *Adding feature for creation of ptl rpm along with pbs rpms, user can also use make install to now install ptl*

#### Affected Platform(s)
* *RHEL(should work for other platforms, currently not tested)*

#### Cause / Analysis / Design
* *Link to external design document **[PTL package design](https://pbspro.atlassian.net/wiki/spaces/PD/pages/672006148/PTL+package+design)***
* *Discussion on the Forum, link : **[pbspro community forum](http://community.pbspro.org/t/proposing-method-for-ptl-packaging/1168/17)***

#### Solution Description
* *Added changes in various makefiles and configure file to add ptl files in a package.*
 * *Added files for profile.d toset ptl paths*
* *Modified spec file to support creation of ptl rpm*
#### Testing logs/output
* 

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [ ] I have attached **test logs to this pull request** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

**Comment by sandisamp** [comment]
Cadacy is failing beceause of ptl.sh which is a custom.sh file which goes in /etc/profile.d
It does not require a shebang line however codacy requires one.

File: INSTALL [hunk 1]
```diff
@@ -75,6 +75,10 @@ How to install PBS Pro using the configure script.
 
     ./configure --prefix=/opt/pbs --libexecdir=/opt/pbs/libexec
 
+  If ptl needs to be installed along with pbspro use the option
```
Comment (mike0042) [inline INSTALL]: PTL is an acronym. Please capitalize it.

File: INSTALL [hunk 2]
```diff
@@ -191,6 +195,11 @@ Note 4: You need to use a POSIX (or nearly POSIX) make. GNU make
   is not guaranteed to work. Don't be surprised if the first thing
   that make does is call configure again.
 
+Note 5: Ptl gets installed in the parent directory of where pbs is installed.
```
Comment (mike0042) [inline INSTALL]: PTL is an acronym. Please capitalize it.
Also, it may be better to say PTL uses the same prefix as PBS.

File: configure.ac [hunk 3]
```diff
@@ -56,6 +56,16 @@ AS_CASE([$os_id],
...
 AC_USE_SYSTEM_EXTENSIONS
 
+AM_PATH_PYTHON([2.7],, [:])
+AM_CONDITIONAL([HAVE_PYTHON], [ test "$PYTHON" != : ])
+
+# check for ptl enable, disabled by default
+AC_ARG_ENABLE([ptl],
+    AS_HELP_STRING([--enable-ptl], [Enable package creation for ptl]))
+
+AS_IF([test "x$enable_ptl" = "xyes"], [AC_MSG_RESULT([checking for ptl...yes])], [AC_MSG_RESULT([checking for ptl...no])])
```
Comment (mike0042) [inline configure.ac]: Please capitalize PTL and add one space after each ellipsis.

File: configure.ac [hunk 4]
```diff
@@ -56,6 +56,16 @@ AS_CASE([$os_id],
...
+AM_CONDITIONAL([ENABLEPTL],[test "x${enable_ptl}" = "xyes"])
```
Comment (mike0042) [inline configure.ac]: Please add a space after the comma.

File: pbspro.spec.in [hunk 5]
```diff
@@ -140,6 +146,29 @@ job scheduler designed to improve productivity, optimize
...
+%endif
+Requires: python-defusedxml
+
+%description %{pbs_ptl}
+PBS Professional® is a fast, powerful workload manager and
+job scheduler designed to improve productivity, optimize
+utilization & efficiency, and simplify administration for
+HPC clusters, clouds and supercomputers.
+
+This package is intended for testing PBS.
```
Comment (mike0042) [inline pbspro.spec.in]: Remove the PBS Professional description and replace the last line with:

PBS Test Lab is a test harness and test suite intended to validate the
functionality of PBS Professional®.

File: test/fw/ptl.csh [hunk 6]
```diff
@@ -0,0 +1,45 @@
+
```
Comment (mike0042) [inline test/fw/ptl.csh]: Add a shebang or Codacy will complain.

File: test/fw/ptl.sh [hunk 7]
```diff
@@ -0,0 +1,45 @@
+
```
Comment (mike0042) [inline test/fw/ptl.sh]: Add a shebang or Codacy will complain.

File: INSTALL [hunk 8]
```diff
@@ -195,8 +195,8 @@ Note 4: You need to use a POSIX (or nearly POSIX) make. GNU make
   is not guaranteed to work. Don't be surprised if the first thing
   that make does is call configure again.
 
-Note 5: Ptl gets installed in the parent directory of where pbs is installed.
-  For example if you have given install prefix=/opt/pbs, then you can find ptl 
+Note 5: PTL gets installed in the parent directory of where pbs is installed.
```
Comment (mike0042) [inline INSTALL]: Sorry I missed this the first time... please change pbs to PBS Professional.

File: pbspro.spec.in (hunk 5 above)
Comment (bhroam) [reply pbspro.spec.in]: PTL isn't just a test harness.  It is a python interface to PBS.  I don't think we should have its explanation as just about testing.

File: configure.ac [hunk 9]
```diff
@@ -56,6 +56,16 @@ AS_CASE([$os_id],
 AM_INIT_AUTOMAKE(am_init_string)
 AC_USE_SYSTEM_EXTENSIONS
 
+AM_PATH_PYTHON([2.7],, [:])
```
Comment (kjakkali) [inline configure.ac]: Create M4 file 'enable_ptl'.

File: pbspro.spec [hunk 10]
```diff
@@ -140,6 +146,25 @@ job scheduler designed to improve productivity, optimize
 utilization & efficiency, and simplify administration for
 HPC clusters, clouds and supercomputers.
 
+%if %{with ptl}
+%package %{pbs_ptl}
+Summary: PBS ptl suite for testing
```
Comment (kjakkali) [inline pbspro.spec]: Update summary to 'PBS Test Lab for testing'

File: pbspro.spec [hunk 11]
```diff
@@ -322,6 +350,11 @@ else
         install -D %{pbs_prefix}/libexec/pbs_init.d /etc/init.d/pbs
 fi
 
+%if %{with ptl}
```
Comment (kjakkali) [inline pbspro.spec]: You don't need post block.

File: pbspro.spec [hunk 12]
```diff
@@ -417,6 +450,19 @@ fi
 %exclude %{pbs_prefix}/unsupported/*.pyc
 %exclude %{pbs_prefix}/unsupported/*.pyo
 
+%if %{with ptl}
+%files %{pbs_ptl}
+%defattr(-,root,root, -)
+%dir %{ptl_prefix}
+%{ptl_prefix}/*
+%exclude %{ptl_prefix}/setup.pyo
```
Comment (kjakkali) [inline pbspro.spec]: Don't exclude any PTL files

File: test/fw/Makefile.am [hunk 13]
```diff
@@ -39,6 +38,80 @@
...
+ptl_libdir = $(ptl_prefixdir)/lib/python$(PYTHON_VERSION)/site-packages/ptl
+ptl_bindir = $(ptl_prefixdir)/bin
+ptl_docdir = $(ptl_prefixdir)/doc
+sysprofiledir = /etc/profile.d
+
+dist_sysprofile_DATA = \
+	ptl.csh \
+	ptl.sh
+
+dist_ptl_bin_SCRIPTS= \
```
Comment (kjakkali) [inline test/fw/Makefile.am]: Try using wildcards instead of listing every files.

File: test/fw/ptl.csh [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+#
+# Use of Altair’s trademarks, including but not limited to "PBS™",
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+# trademark licensing policies.
+
+# This file will set path variables in case of ptl installation
+
+ptl_prefix_lib=$( rpm -ql pbspro-ptl | grep -m 1 lib$ )
+ptl_prefix_bin=$( rpm -ql pbspro-ptl | grep -m 1 bin$ )
+python_dir=$( ls ${ptl_prefix_lib} )
```
Comment (kjakkali) [inline test/fw/ptl.csh]: use /bin/ls instead of ls . use -1 option to ls

File: test/fw/ptl.csh [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+prefix=$( printf "%s\n%s\n" "$ptl_prefix_bin" "$ptl_prefix_lib" | sed -e 'N;s/^\(.*\).*\n\1.*$/\1/' )
```
Comment (kjakkali) [inline test/fw/ptl.csh]: Use dirname command

File: test/fw/ptl.csh [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+
+setenv PATH=${PATH}:${prefix}bin/
+setenv PYTHONPATH=$PYTHONPATH:${prefix}lib/${python_dir}/site-packages/
```
Comment (kjakkali) [inline test/fw/ptl.csh]: Add new path at the start.

File: test/fw/ptl.sh [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+# This file will set path variables in case of ptl installation
+
+ptl_prefix_lib=$( rpm -ql pbspro-ptl | grep -m 1 lib$ )
+ptl_prefix_bin=$( rpm -ql pbspro-ptl | grep -m 1 bin$ )
+python_dir=$( ls ${ptl_prefix_lib} )
+prefix=$( printf "%s\n%s\n" "$ptl_prefix_bin" "$ptl_prefix_lib" | sed -e 'N;s/^\(.*\).*\n\1.*$/\1/' )
+
+export PATH=${PATH}:${prefix}bin/
+export PYTHONPATH=$PYTHONPATH:${prefix}lib/${python_dir}/site-packages/
```
Comment (kjakkali) [inline test/fw/ptl.sh]: Add new path at the start.

File: test/tests/Makefile.am [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+if ENABLEPTL
+ptl_prefixdir = $(prefix)/../ptl
+ptl_testsdir = $(ptl_prefixdir)/tests
+
+dist_ptl_tests_PYTHON = \
+    __init__.py \
+    pbs_smoketest.py 
+
+ptl_testfunctionaldir = $(ptl_testsdir)/functional
+dist_ptl_testfunctional_DATA = \
```
Comment (kjakkali) [inline test/tests/Makefile.am]: Try using wildcards

File: test/fw/Makefile.am [hunk 19]
```diff
@@ -39,6 +38,80 @@
...
+	bin/pbs_py_spawn \
+	bin/pbs_snapshot \
+	bin/pbs_stat \
+	bin/pbs_swigify
+
+ptl_lib_libdir = $(ptl_libdir)/lib
+
+dist_ptl_lib_PYTHON= ptl/__init__.py
+
+dist_ptl_doc_DATA = \
```
Comment (kjakkali) [inline test/fw/Makefile.am]: You don't need include doc dir.

File: test/fw/Makefile.am [hunk 20]
```diff
@@ -88,3 +161,4 @@ install-data-hook:
 uninstall-hook:
 	cd $(DESTDIR)$(ptlbindir) && \
 	rm -f pbs_stat.py pbs_loganalyzer.py pbs_snapshot.py
+
```
Comment (kjakkali) [inline test/fw/Makefile.am]: remove blank line

File: test/tests/Makefile.am [hunk 21]
```diff
@@ -0,0 +1,198 @@
+
```
Comment (kjakkali) [inline test/tests/Makefile.am]: remove blank line

File: configure.ac [hunk 22]
```diff
@@ -56,6 +56,9 @@ AS_CASE([$os_id],
 AM_INIT_AUTOMAKE(am_init_string)
 AC_USE_SYSTEM_EXTENSIONS
 
+AM_PATH_PYTHON(,, [:])
```
Comment (hirenvadalia) [inline configure.ac]: Why we need this change?

File: m4/enable_ptl.m4 [hunk 23]
```diff
@@ -0,0 +1,13 @@
+AC_DEFUN([PBS_AC_ENABLE_PTL],
```
Comment (hirenvadalia) [inline m4/enable_ptl.m4]: Please add license header at top and newline at the end.

File: test/tests/Makefile.am [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+#
+# Use of Altair’s trademarks, including but not limited to "PBS™",
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+# trademark licensing policies.
+#
+if ENABLEPTL
+ptl_prefixdir = $(prefix)/../ptl
+ptl_testsdir = $(ptl_prefixdir)/tests
+
+dist_ptl_tests_PYTHON = \
```
Comment (hirenvadalia) [inline test/tests/Makefile.am]: Use wildcard please.

File: test/fw/Makefile.am [hunk 25]
```diff
@@ -36,6 +35,65 @@
...
+	bin/pbs_snapshot \
+	bin/pbs_stat \
+	bin/pbs_swigify
+
+ptl_lib_libdir = $(ptl_libdir)/lib
+
+dist_ptl_lib_PYTHON= ptl/__init__.py
+
+dist_ptl_lib_lib_PYTHON = \
+	ptl/lib/pbs_api_to_cli.py \
```
Comment (hirenvadalia) [inline test/fw/Makefile.am]: Use wildcard in this file also.

File: configure.ac [hunk 26]
```diff
@@ -336,6 +339,7 @@ PBS_AC_DISABLE_SHELL_PIPE
 PBS_AC_DISABLE_SYSLOG
 PBS_AC_ENABLE_ALPS
 PBS_AC_ENABLE_CPUSET
+PBS_AC_ENABLE_PTL
```
Comment (saritakh) [inline configure.ac]: Move this line after next line

File: test/fw/ptl.csh [hunk 27]
```diff
@@ -0,0 +1,45 @@
...
+# trademark licensing policies.
+
+# This file will set path variables in case of ptl installation
+
+ptl_prefix_lib=$( rpm -ql pbspro-ptl | grep -m 1 lib$ )
+python_dir=$( /bin/ls -1 ${ptl_prefix_lib} )
+prefix=$( dirname ${ptl_prefix_lib} )
+
+setenv PATH=${prefix}/bin/:${PATH}
+setenv PYTHONPATH=${prefix}/lib/${python_dir}/site-packages/:$PYTHONPATH
```
Comment (hirenvadalia) [inline test/fw/ptl.csh]: missing new line

**Comment by sandisamp** [comment]
@mike0042 Do you recommend any further changes?

**Comment by sandisamp** [comment]
@mike0042 The doc folder is not part of the rpm package so it should not cause any hurdles. Also regarding the EPEL repo, I've already mentioned it in the EDD for this ptl packaging.  Should I apply your suggested changes anyway?

**Comment by mike0042** [comment]
I think you need to document the requirement for installing pip. It's not something that can easily be added to the spec file, but also something that could frustrate a person trying to setup PTL.

**Comment by mike0042** [comment]
Thanks @sandisamp, it all looks good to me.

**Comment by mike0042** [comment]
@PBSPro/pbspro-maintainers : This PR appears to have broken the build regardless of whether --enable-ptl is specified for configure. I ran my build on the same CentOS 7 VM I normally use.
```
make[2]: Entering directory `/home/mike/pbsbuild/pbspro/test'
 (cd fw && make  top_distdir=../../pbspro-19.1.5 distdir=../../pbspro-19.1.5/test/fw \
     am__remove_distdir=: am__skip_length_check=: am__skip_mode_fix=: distdir)
make[3]: Entering directory `/home/mike/pbsbuild/pbspro/test/fw'
make[3]: *** No rule to make target `../../repo/test/fw/ptl/*.py', needed by `distdir'.  Stop.
make[3]: Leaving directory `/home/mike/pbsbuild/pbspro/test/fw'
make[2]: *** [distdir] Error 1
make[2]: Leaving directory `/home/mike/pbsbuild/pbspro/test'
make[1]: *** [distdir] Error 1
make[1]: Leaving directory `/home/mike/pbsbuild/pbspro'
make: *** [dist] Error 2
```

**Comment by mike0042** [comment]
GNU automake recommends avoiding wildcards:
https://www.gnu.org/software/automake/manual/html_node/Wildcards.html#Wildcards

**Pull request merged**